File: src/org/jetbrains/plugins/scala/debugger/ScalaJVMNameMapper.scala
package org.jetbrains.plugins.scala
package debugger

import com.intellij.psi.PsiClass
import com.intellij.openapi.application.ApplicationManager
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.{ScTrait, ScObject, ScTemplateDefinition}
import com.intellij.openapi.util.Computable
import org.jetbrains.annotations.NotNull
import com.intellij.debugger.NameMapper

/**
*@author ilyas
*/
class ScalaJVMNameMapper extends NameMapper {
  def getQualifiedName(@NotNull clazz: PsiClass): String = {
    ApplicationManager.getApplication.runReadAction(new Computable[String] {
      def compute: String = {
        clazz match {
          case obj: ScObject => obj.qualifiedName + "$"
          case tr: ScTrait => tr.qualifiedName
          case templDef: ScTemplateDefinition => templDef.qualifiedName
          case psiClass => psiClass.getQualifiedName
        }
      }
    })
  }
}

File: src/org/jetbrains/plugins/scala/debugger/ScalaJVMNameMapper.java
package org.jetbrains.plugins.scala.debugger;

import com.intellij.debugger.NameMapper;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.util.Computable;
import com.intellij.psi.PsiClass;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.ScObject;
import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.ScTrait;
import org.jetbrains.plugins.scala.util.ScalaUtils;

/**
 * @author ilyas
 */
public class ScalaJVMNameMapper implements NameMapper {

  public String getQualifiedName(@NotNull final PsiClass clazz) {
    String qualifiedName = ApplicationManager.getApplication().runReadAction(new Computable<String>() {public String compute() {
      return clazz.getQualifiedName();
    }
    });
    if (clazz instanceof ScObject) return qualifiedName + "$";
    if (clazz instanceof ScTrait) return qualifiedName + "$class";
    return qualifiedName;
  }
}

class StringContainer {
  public StringContainer(String s) {
    this.s = s;
  }
  private String s;

  public String getS() {
    return s;
  }

  public void setS(String s) {
    this.s = s;
  }
}
